// @superfluid-finance/sdk-core/dist/main/subgraph/entities/accountTokenSnapshot/accountTokenSnapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountTokenSnapshotQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const accountTokenSnapshots_generated_1 = require("./accountTokenSnapshots.generated");
class AccountTokenSnapshotQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["account"],
            tokenKeys: ["token"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.token],
            accounts: [result.account],
        });
        this.mapFromSubgraphResponse = (response) => response.accountTokenSnapshots.map((x) => ({
            ...x,
            account: x.account.id,
            token: x.token.id,
            tokenSymbol: x.token.symbol,
            maybeCriticalAtTimestamp: x.maybeCriticalAtTimestamp != null
                ? Number(x.maybeCriticalAtTimestamp)
                : null,
            updatedAtBlockNumber: Number(x.updatedAtBlockNumber),
            updatedAtTimestamp: Number(x.updatedAtTimestamp),
        }));
        this.requestDocument = accountTokenSnapshots_generated_1.AccountTokenSnapshotsDocument;
    }
}
exports.AccountTokenSnapshotQueryHandler = AccountTokenSnapshotQueryHandler;
