// @superfluid-finance/sdk-core/dist/module/typechain/factories/SuperfluidLoader__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "contract IResolver",
                name: "resolver",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "releaseVersion",
                type: "string",
            },
        ],
        name: "loadFramework",
        outputs: [
            {
                components: [
                    {
                        internalType: "contract ISuperfluid",
                        name: "superfluid",
                        type: "address",
                    },
                    {
                        internalType: "contract ISuperTokenFactory",
                        name: "superTokenFactory",
                        type: "address",
                    },
                    {
                        internalType: "contract ISuperAgreement",
                        name: "agreementCFAv1",
                        type: "address",
                    },
                    {
                        internalType: "contract ISuperAgreement",
                        name: "agreementIDAv1",
                        type: "address",
                    },
                ],
                internalType: "struct SuperfluidLoader.Framework",
                name: "result",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class SuperfluidLoader__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
SuperfluidLoader__factory.abi = _abi;
