// @superfluid-finance/sdk-core/dist/main/subgraph/entities/indexSubscription/indexSubscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexSubscriptionQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const indexSubscriptions_generated_1 = require("./indexSubscriptions.generated");
class IndexSubscriptionQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["subscriber"],
            tokenKeys: [],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.token],
            accounts: [result.subscriber, result.publisher],
        });
        this.mapFromSubgraphResponse = (response) => response.indexSubscriptions.map((x) => ({
            ...x,
            subscriber: x.subscriber.id,
            createdAtTimestamp: Number(x.createdAtTimestamp),
            createdAtBlockNumber: Number(x.createdAtBlockNumber),
            updatedAtTimestamp: Number(x.updatedAtTimestamp),
            updatedAtBlockNumber: Number(x.updatedAtBlockNumber),
            index: x.index.id,
            indexValueCurrent: x.index.indexValue,
            indexTotalUnits: x.index.totalUnits,
            token: x.index.token.id,
            tokenSymbol: x.index.token.symbol,
            publisher: x.index.publisher.id,
        }));
        this.requestDocument = indexSubscriptions_generated_1.IndexSubscriptionsDocument;
    }
}
exports.IndexSubscriptionQueryHandler = IndexSubscriptionQueryHandler;
