// @superfluid-finance/sdk-core/dist/main/subgraph/entities/token/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const tokens_generated_1 = require("./tokens.generated");
class TokenQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: [],
            tokenKeys: ["id", "underlyingToken", "underlyingAddress"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.underlyingAddress, result.id],
            accounts: [],
        });
        this.mapFromSubgraphResponse = (response) => response.tokens.map((x) => ({
            ...x,
            createdAtBlockNumber: Number(x.createdAtBlockNumber),
            createdAtTimestamp: Number(x.createdAtTimestamp),
        }));
        this.requestDocument = tokens_generated_1.TokensDocument;
    }
}
exports.TokenQueryHandler = TokenQueryHandler;
