// @superfluid-finance/sdk-core/dist/module/subgraph/entities/accountTokenSnapshotLog/accountTokenSnapshotLog.js
import { SubgraphQueryHandler, } from "../../subgraphQueryHandler";
import { AccountTokenSnapshotLogsDocument, } from "./accountTokenSnapshotLogs.generated";
export class AccountTokenSnapshotLogQueryHandler extends SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["account"],
            tokenKeys: ["token"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.token],
            accounts: [result.account],
        });
        this.mapFromSubgraphResponse = (response) => response.accountTokenSnapshotLogs.map((x) => ({
            ...x,
            account: x.account.id,
            token: x.token.id,
            tokenSymbol: x.token.symbol,
            maybeCriticalAtTimestamp: x.maybeCriticalAtTimestamp != null
                ? Number(x.maybeCriticalAtTimestamp)
                : null,
            blockNumber: Number(x.blockNumber),
            timestamp: Number(x.timestamp),
            order: Number(x.order),
            logIndex: Number(x.logIndex),
        }));
        this.requestDocument = AccountTokenSnapshotLogsDocument;
    }
}
