// @superfluid-finance/sdk-core/dist/main/subgraph/events/flowUpdatedEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowUpdatedEventQueryHandler = void 0;
const mapGetAllEventsQueryEvents_1 = require("../../mapGetAllEventsQueryEvents");
const subgraphQueryHandler_1 = require("../subgraphQueryHandler");
const events_generated_1 = require("./events.generated");
class FlowUpdatedEventQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["sender", "receiver", "flowOperator"],
            tokenKeys: ["token"],
        });
        this.requestDocument = events_generated_1.FlowUpdatedEventsDocument;
    }
    getRelevantAddressesFromResultCore(result) {
        return {
            accounts: [result.sender, result.receiver],
            tokens: [result.token],
        };
    }
    mapFromSubgraphResponse(response) {
        return (0, mapGetAllEventsQueryEvents_1.mapGetAllEventsQueryEvents)(response.flowUpdatedEvents);
    }
}
exports.FlowUpdatedEventQueryHandler = FlowUpdatedEventQueryHandler;
