// @superfluid-finance/sdk-core/dist/main/subgraph/events/indexUpdatedEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexUpdatedEventQueryHandler = void 0;
const mapGetAllEventsQueryEvents_1 = require("../../mapGetAllEventsQueryEvents");
const subgraphQueryHandler_1 = require("../subgraphQueryHandler");
const events_generated_1 = require("./events.generated");
class IndexUpdatedEventQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["publisher"],
            tokenKeys: ["token"],
        });
        this.requestDocument = events_generated_1.IndexUpdatedEventsDocument;
    }
    getRelevantAddressesFromResultCore(result) {
        return {
            accounts: [result.publisher],
            tokens: [result.token],
        };
    }
    mapFromSubgraphResponse(response) {
        return (0, mapGetAllEventsQueryEvents_1.mapGetAllEventsQueryEvents)(response.indexUpdatedEvents);
    }
}
exports.IndexUpdatedEventQueryHandler = IndexUpdatedEventQueryHandler;
