// @superfluid-finance/sdk-core/dist/main/subgraph/events/subscriptionUnitsUpdatedEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionUnitsUpdatedEventQueryHandler = void 0;
const mapGetAllEventsQueryEvents_1 = require("../../mapGetAllEventsQueryEvents");
const subgraphQueryHandler_1 = require("../subgraphQueryHandler");
const events_generated_1 = require("./events.generated");
class SubscriptionUnitsUpdatedEventQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["publisher", "subscriber"],
            tokenKeys: ["token"],
        });
        this.requestDocument = events_generated_1.SubscriptionUnitsUpdatedEventsDocument;
    }
    getRelevantAddressesFromResultCore(result) {
        return {
            accounts: [result.publisher, result.subscriber],
            tokens: [result.token],
        };
    }
    mapFromSubgraphResponse(response) {
        return (0, mapGetAllEventsQueryEvents_1.mapGetAllEventsQueryEvents)(response.subscriptionUnitsUpdatedEvents);
    }
}
exports.SubscriptionUnitsUpdatedEventQueryHandler = SubscriptionUnitsUpdatedEventQueryHandler;
