// @superfluid-finance/sdk-core/dist/main/subgraph/events/transferEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferEventQueryHandler = void 0;
const mapGetAllEventsQueryEvents_1 = require("../../mapGetAllEventsQueryEvents");
const subgraphQueryHandler_1 = require("../subgraphQueryHandler");
const events_generated_1 = require("./events.generated");
class TransferEventQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["from", "to"],
            tokenKeys: ["token"],
        });
        this.requestDocument = events_generated_1.TransferEventsDocument;
    }
    getRelevantAddressesFromResultCore(result) {
        return {
            accounts: [result.from, result.to],
            tokens: [result.token],
        };
    }
    mapFromSubgraphResponse(response) {
        return (0, mapGetAllEventsQueryEvents_1.mapGetAllEventsQueryEvents)(response.transferEvents);
    }
}
exports.TransferEventQueryHandler = TransferEventQueryHandler;
