// @superfluid-finance/sdk-core/dist/main/typechain/factories/SuperfluidLoader__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuperfluidLoader__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "contract IResolver",
                name: "resolver",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "releaseVersion",
                type: "string",
            },
        ],
        name: "loadFramework",
        outputs: [
            {
                components: [
                    {
                        internalType: "contract ISuperfluid",
                        name: "superfluid",
                        type: "address",
                    },
                    {
                        internalType: "contract ISuperTokenFactory",
                        name: "superTokenFactory",
                        type: "address",
                    },
                    {
                        internalType: "contract ISuperAgreement",
                        name: "agreementCFAv1",
                        type: "address",
                    },
                    {
                        internalType: "contract ISuperAgreement",
                        name: "agreementIDAv1",
                        type: "address",
                    },
                ],
                internalType: "struct SuperfluidLoader.Framework",
                name: "result",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class SuperfluidLoader__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.SuperfluidLoader__factory = SuperfluidLoader__factory;
SuperfluidLoader__factory.abi = _abi;
