// @superfluid-finance/sdk-core/dist/main/subgraph/entities/accountTokenSnapshotLog/accountTokenSnapshotLog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountTokenSnapshotLogQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const accountTokenSnapshotLogs_generated_1 = require("./accountTokenSnapshotLogs.generated");
class AccountTokenSnapshotLogQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["account"],
            tokenKeys: ["token"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.token],
            accounts: [result.account],
        });
        this.mapFromSubgraphResponse = (response) => response.accountTokenSnapshotLogs.map((x) => ({
            ...x,
            account: x.account.id,
            token: x.token.id,
            tokenSymbol: x.token.symbol,
            maybeCriticalAtTimestamp: x.maybeCriticalAtTimestamp != null
                ? Number(x.maybeCriticalAtTimestamp)
                : null,
            blockNumber: Number(x.blockNumber),
            timestamp: Number(x.timestamp),
            order: Number(x.order),
            logIndex: Number(x.logIndex),
        }));
        this.requestDocument = accountTokenSnapshotLogs_generated_1.AccountTokenSnapshotLogsDocument;
    }
}
exports.AccountTokenSnapshotLogQueryHandler = AccountTokenSnapshotLogQueryHandler;
