// @superfluid-finance/sdk-core/dist/module/subgraph/entities/stream/stream.js
import { SubgraphQueryHandler, } from "../../subgraphQueryHandler";
import { StreamsDocument, } from "./streams.generated";
export class StreamQueryHandler extends SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["sender", "receiver"],
            tokenKeys: ["token"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.token],
            accounts: [result.sender, result.receiver],
        });
        this.mapFromSubgraphResponse = (response) => response.streams.map((x) => ({
            ...x,
            createdAtTimestamp: Number(x.createdAtTimestamp),
            createdAtBlockNumber: Number(x.createdAtBlockNumber),
            updatedAtTimestamp: Number(x.updatedAtTimestamp),
            updatedAtBlockNumber: Number(x.updatedAtBlockNumber),
            receiver: x.receiver.id,
            token: x.token.id,
            tokenSymbol: x.token.symbol,
            sender: x.sender.id,
        }));
        this.requestDocument = StreamsDocument;
    }
}
