// @superfluid-finance/sdk-core/dist/main/subgraph/entities/account/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const accounts_generated_1 = require("./accounts.generated");
class AccountQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: ["id"],
            tokenKeys: [],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            accounts: [result.id],
            tokens: [],
        });
        this.mapFromSubgraphResponse = (response) => response.accounts.map((x) => ({
            ...x,
            createdAtTimestamp: Number(x.createdAtTimestamp),
            createdAtBlockNumber: Number(x.createdAtBlockNumber),
            updatedAtTimestamp: Number(x.updatedAtTimestamp),
            updatedAtBlockNumber: Number(x.updatedAtBlockNumber),
        }));
        this.requestDocument = accounts_generated_1.AccountsDocument;
    }
}
exports.AccountQueryHandler = AccountQueryHandler;
