// @superfluid-finance/sdk-core/dist/main/Host.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
const Operation_1 = __importDefault(require("./Operation"));
const Superfluid_json_1 = __importDefault(require("./abi/Superfluid.json"));
/**
 * Host Helper Class
 * @description A helper class which can be used as a standalone class to populate call agreement transactions.
 */
class Host {
    constructor(hostAddress) {
        /**
         * Creates an Operation of the `callAgreement` function on the host contract.
         * @param agreementAddress the agreement address (cfa or ida address)
         * @param callData the encoded callData for the function
         * @param userData any additional user data
         * @param overrides ethers overrides object for more control over the transaction sent.
         * @returns {Operation} an `Operation` class
         */
        this.populateCallAgreementTxnAndReturnOperation = (agreementAddress, callData, userData, overrides) => {
            const txn = this.contract.populateTransaction.callAgreement(agreementAddress, callData, userData || "0x", overrides || {});
            return new Operation_1.default(txn, "SUPERFLUID_CALL_AGREEMENT");
        };
        this.contract = new ethers_1.ethers.Contract(hostAddress, Superfluid_json_1.default.abi);
    }
}
exports.default = Host;
