// @superfluid-finance/sdk-core/dist/main/mapGetAllEventsQueryEvents.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapGetAllEventsQueryEvents = void 0;
const events = __importStar(require("./events"));
const utils_1 = require("./utils");
const mapGetAllEventsQueryEvents = (queryEvents) => {
    return queryEvents.map((x) => {
        switch (x.__typename) {
            case "AppRegisteredEvent":
                return (0, utils_1.typeGuard)({
                    name: "AppRegistered",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    app: x.app,
                });
            case "AgreementClassRegisteredEvent":
                return (0, utils_1.typeGuard)({
                    name: "AgreementClassRegistered",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    agreementType: x.agreementType,
                    code: x.code,
                });
            case "AgreementClassUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "AgreementClassUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    agreementType: x.agreementType,
                    code: x.code,
                });
            case "AgreementLiquidatedByEvent":
                return (0, utils_1.typeGuard)({
                    name: "AgreementLiquidatedBy",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    agreementClass: x.agreementClass,
                    agreementId: x.agreementId,
                    bailoutAmount: x.bailoutAmount,
                    bondAccount: x.bondAccount,
                    liquidatorAccount: x.liquidatorAccount,
                    penaltyAccount: x.penaltyAccount,
                    rewardAmount: x.rewardAmount,
                    token: x.token,
                });
            case "AgreementLiquidatedV2Event":
                return (0, utils_1.typeGuard)({
                    name: "AgreementLiquidatedV2",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    liquidatorAccount: x.liquidatorAccount,
                    agreementClass: x.agreementClass,
                    agreementId: x.agreementId,
                    targetAccount: x.targetAccount,
                    rewardAmountReceiver: x.rewardAmountReceiver,
                    rewardAmount: x.rewardAmount,
                    targetAccountBalanceDelta: x.targetAccountBalanceDelta,
                    version: x.version,
                    liquidationType: x.liquidationType,
                    // TO BE DEPRECATED in v2 endpoint - use rewardAmountReceiver instead
                    rewardAccount: x.rewardAccount,
                });
            case "BurnedEvent":
                return (0, utils_1.typeGuard)({
                    name: "Burned",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    amount: x.amount,
                    data: x.data,
                    from: x.from,
                    operator: x.operator,
                    token: x.token,
                    operatorData: x.operatorData,
                });
            case "CFAv1LiquidationPeriodChangedEvent":
                return (0, utils_1.typeGuard)({
                    name: "CFAv1LiquidationPeriodChanged",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    host: x.host,
                    governanceAddress: x.governanceAddress,
                    isKeySet: x.isKeySet,
                    liquidationPeriod: Number(x.liquidationPeriod),
                    superToken: x.superToken,
                });
            case "ConfigChangedEvent":
                return (0, utils_1.typeGuard)({
                    name: "ConfigChanged",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    host: x.host,
                    governanceAddress: x.governanceAddress,
                    isKeySet: x.isKeySet,
                    key: x.key,
                    superToken: x.superToken,
                    value: x.value,
                });
            case "CustomSuperTokenCreatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "CustomSuperTokenCreated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                });
            case "FlowUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "FlowUpdated",
                    type: mapNumberToFlowUpdateType(x.type),
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    flowRate: x.flowRate,
                    receiver: x.receiver,
                    sender: x.sender,
                    flowOperator: x.flowOperator,
                    deposit: x.deposit,
                });
            case "FlowOperatorUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "FlowOperatorUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    sender: x.sender,
                    permissions: x.permissions,
                    flowRateAllowance: x.flowRateAllowance,
                });
            case "GovernanceReplacedEvent":
                return (0, utils_1.typeGuard)({
                    name: "GovernanceReplaced",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    oldGovernance: x.oldGovernance,
                    newGovernance: x.newGovernance,
                });
            case "IndexCreatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "IndexCreated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    indexId: x.indexId,
                    publisher: x.publisher,
                    userData: x.userData,
                });
            case "IndexDistributionClaimedEvent":
                return (0, utils_1.typeGuard)({
                    name: "IndexDistributionClaimed",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    amount: x.amount,
                    indexId: x.indexId,
                    publisher: x.publisher,
                    subscriber: x.subscriber,
                });
            case "IndexSubscribedEvent":
                return (0, utils_1.typeGuard)({
                    name: "IndexSubscribed",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    indexId: x.indexId,
                    publisher: x.publisher,
                    subscriber: x.subscriber,
                });
            case "IndexUnitsUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "IndexUnitsUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    indexId: x.indexId,
                    publisher: x.publisher,
                    subscriber: x.subscriber,
                    units: x.units,
                    oldUnits: x.oldUnits,
                    userData: x.userData,
                });
            case "IndexUnsubscribedEvent":
                return (0, utils_1.typeGuard)({
                    name: "IndexUnsubscribed",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    indexId: x.indexId,
                    publisher: x.publisher,
                    subscriber: x.subscriber,
                    userData: x.userData,
                });
            case "IndexUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "IndexUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    indexId: x.indexId,
                    newIndexValue: x.newIndexValue,
                    oldIndexValue: x.oldIndexValue,
                    publisher: x.publisher,
                    totalUnitsApproved: x.totalUnitsApproved,
                    totalUnitsPending: x.totalUnitsPending,
                    userData: x.userData,
                });
            case "JailEvent":
                return (0, utils_1.typeGuard)({
                    name: "Jail",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    app: x.app,
                    reason: x.reason,
                });
            case "MintedEvent":
                return (0, utils_1.typeGuard)({
                    name: "Minted",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    operator: x.operator,
                    to: x.to,
                    amount: x.amount,
                    data: x.data,
                    token: x.token,
                    operatorData: x.operatorData,
                });
            case "RewardAddressChangedEvent":
                return (0, utils_1.typeGuard)({
                    name: "RewardAddressChanged",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    host: x.host,
                    governanceAddress: x.governanceAddress,
                    superToken: x.superToken,
                    isKeySet: x.isKeySet,
                    rewardAddress: x.rewardAddress,
                });
            case "RoleAdminChangedEvent":
                return (0, utils_1.typeGuard)({
                    name: "RoleAdminChanged",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    role: x.role,
                    previousAdminRole: x.previousAdminRole,
                    newAdminRole: x.newAdminRole,
                });
            case "RoleGrantedEvent":
                return (0, utils_1.typeGuard)({
                    name: "RoleGranted",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    role: x.role,
                    account: x.account,
                    sender: x.sender,
                });
            case "RoleRevokedEvent":
                return (0, utils_1.typeGuard)({
                    name: "RoleRevoked",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    role: x.role,
                    account: x.account,
                    sender: x.sender,
                });
            case "SentEvent":
                return (0, utils_1.typeGuard)({
                    name: "Sent",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    operator: x.operator,
                    to: x.to,
                    amount: x.amount,
                    data: x.data,
                    token: x.token,
                    operatorData: x.operatorData,
                });
            case "SubscriptionApprovedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SubscriptionApproved",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    subscriber: x.subscriber,
                    publisher: x.publisher,
                    indexId: x.indexId,
                    userData: x.userData,
                    subscription: x.subscription.id,
                });
            case "SubscriptionDistributionClaimedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SubscriptionDistributionClaimed",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    subscriber: x.subscriber,
                    publisher: x.publisher,
                    indexId: x.indexId,
                    amount: x.amount,
                    subscription: x.subscription.id,
                });
            case "SubscriptionRevokedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SubscriptionRevoked",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    subscriber: x.subscriber,
                    publisher: x.publisher,
                    indexId: x.indexId,
                    userData: x.userData,
                    subscription: x.subscription.id,
                });
            case "SubscriptionUnitsUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SubscriptionUnitsUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    subscriber: x.subscriber,
                    publisher: x.publisher,
                    indexId: x.indexId,
                    units: x.units,
                    userData: x.userData,
                    subscription: x.subscription.id,
                });
            case "SuperTokenCreatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SuperTokenCreated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                });
            case "SuperTokenFactoryUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SuperTokenFactoryUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    newFactory: x.newFactory,
                });
            case "SuperTokenLogicCreatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SuperTokenLogicCreated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    tokenLogic: x.tokenLogic,
                });
            case "SuperTokenLogicUpdatedEvent":
                return (0, utils_1.typeGuard)({
                    name: "SuperTokenLogicUpdated",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    token: x.token,
                    code: x.code,
                });
            case "PPPConfigurationChangedEvent":
                return (0, utils_1.typeGuard)({
                    name: "PPPConfigurationChanged",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    host: x.host,
                    governanceAddress: x.governanceAddress,
                    superToken: x.superToken,
                    isKeySet: x.isKeySet,
                    liquidationPeriod: x.liquidationPeriod,
                    patricianPeriod: x.patricianPeriod,
                });
            case "TokenDowngradedEvent":
                return (0, utils_1.typeGuard)({
                    name: "TokenDowngraded",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    amount: x.amount,
                    token: x.token,
                    account: x.account.id,
                });
            case "TokenUpgradedEvent":
                return (0, utils_1.typeGuard)({
                    name: "TokenUpgraded",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    amount: x.amount,
                    token: x.token,
                    account: x.account.id,
                });
            case "TransferEvent":
                return (0, utils_1.typeGuard)({
                    name: "Transfer",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    from: x.from.id,
                    to: x.to.id,
                    token: x.token,
                    value: x.value,
                });
            case "TrustedForwarderChangedEvent":
                return (0, utils_1.typeGuard)({
                    name: "TrustedForwarderChanged",
                    id: x.id,
                    blockNumber: Number(x.blockNumber),
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: Number(x.order),
                    timestamp: Number(x.timestamp),
                    logIndex: Number(x.logIndex),
                    host: x.host,
                    governanceAddress: x.governanceAddress,
                    superToken: x.superToken,
                    isKeySet: x.isKeySet,
                    forwarder: x.forwarder,
                    enabled: x.enabled,
                });
            default:
                console.warn("An unknown event was detected which couldn't be mapped. Please update to the latest version of @superfluid-finance/sdk-core.");
                return (0, utils_1.typeGuard)({
                    name: "_Unknown",
                    id: x.id,
                    blockNumber: x.blockNumber,
                    transactionHash: x.transactionHash,
                    gasPrice: x.gasPrice,
                    order: x.order,
                    timestamp: x.timestamp,
                    logIndex: x.logIndex,
                });
        }
    });
};
exports.mapGetAllEventsQueryEvents = mapGetAllEventsQueryEvents;
const mapNumberToFlowUpdateType = (type) => {
    if (type === 0) {
        return events.FlowUpdateType.Create;
    }
    else if (type === 1) {
        return events.FlowUpdateType.Update;
    }
    else if (type === 2) {
        return events.FlowUpdateType.Terminate;
    }
    else {
        throw Error(`Mapping error: unknown flow update type ([${type}]).`);
    }
};
