// @superfluid-finance/sdk-core/dist/main/subgraph/entities/tokenStatistic/tokenStatistic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenStatisticQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const tokenStatistics_generated_1 = require("./tokenStatistics.generated");
class TokenStatisticQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: [],
            tokenKeys: ["id", "token"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.id],
            accounts: [],
        });
        this.mapFromSubgraphResponse = (response) => response.tokenStatistics.map((x) => ({
            ...x,
            updatedAtBlockNumber: Number(x.updatedAtBlockNumber),
            updatedAtTimestamp: Number(x.updatedAtTimestamp),
        }));
        this.requestDocument = tokenStatistics_generated_1.TokenStatisticsDocument;
    }
}
exports.TokenStatisticQueryHandler = TokenStatisticQueryHandler;
