// @superfluid-finance/sdk-core/dist/main/subgraph/entities/tokenStatisticLog/tokenStatisticLog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenStatisticLogQueryHandler = void 0;
const subgraphQueryHandler_1 = require("../../subgraphQueryHandler");
const tokenStatisticLogs_generated_1 = require("./tokenStatisticLogs.generated");
class TokenStatisticLogQueryHandler extends subgraphQueryHandler_1.SubgraphQueryHandler {
    constructor() {
        super(...arguments);
        this.getAddressFieldKeysFromFilter = () => ({
            accountKeys: [],
            tokenKeys: ["token"],
        });
        this.getRelevantAddressesFromResultCore = (result) => ({
            tokens: [result.token],
            accounts: [],
        });
        this.mapFromSubgraphResponse = (response) => response.tokenStatisticLogs.map((x) => ({
            ...x,
            blockNumber: Number(x.blockNumber),
            timestamp: Number(x.timestamp),
            token: x.token.id,
        }));
        this.requestDocument = tokenStatisticLogs_generated_1.TokenStatisticLogsDocument;
    }
}
exports.TokenStatisticLogQueryHandler = TokenStatisticLogQueryHandler;
